temporalio/client.py

```diff
@@ -1506,6 +1663,295 @@ class WorkflowExecutionStatus(IntEnum):
+            raise RuntimeError("No page fetched yet")
+        return self._current_page
+
+    @property
+    def next_page_token(self) -> Optional[bytes]:
+        """Token for the next page request if any."""
+        return self._next_page_token
+
+    async def fetch_next_page(self, *, page_size: Optional[int] = None) -> None:
+        """Fetch the next page if any.
+
+        Args:
+            page_size: Override the page size this iterator was originally
+                created with.
+        """
+        resp = await self._client.workflow_service.list_workflow_executions(
+            temporalio.api.workflowservice.v1.ListWorkflowExecutionsRequest(
+                namespace=self._client.namespace,
+                page_size=page_size or self._input.page_size,
+                next_page_token=self._next_page_token or b"",
+                query=self._input.query or "",
+            ),
+            retry=True,
+            metadata=self._input.rpc_metadata,
+            timeout=self._input.rpc_timeout,
+        )
+        self._current_page = [
+            WorkflowExecution._from_raw_info(v, self._client.data_converter)
+            for v in resp.executions
+        ]
+        self._current_page_index = 0
+        self._next_page_token = resp.next_page_token or None
+
+    def __aiter__(self) -> WorkflowExecutionAsyncIterator:
+        """Return self as the iterator."""
+        return self
+
+    async def __anext__(self) -> WorkflowExecution:
+        """Get the next execution on this iterator, fetching next page if
+        necessary.
+        """
+        if self._current_page is None:
+            raise RuntimeError("No page fetched yet")
+        while True:
+            # No more left in page?
+            if self._current_page_index >= len(self._current_page):
+                # If there is a next page token, try to get another page and try
+                # again
+                if self._next_page_token is not None:
+                    await self.fetch_next_page()
+                    continue
+                # No more pages means we're done
+                raise StopAsyncIteration
+            # Get current, increment page index, and return
+            ret = self._current_page[self._current_page_index]
+            self._current_page_index += 1
+            return ret
+
+    async def map_histories(
+        self,
+        *,
+        event_filter_type: WorkflowHistoryEventFilterType = WorkflowHistoryEventFilterType.ALL_EVENT,
+        skip_archival: bool = False,
+        rpc_metadata: Mapping[str, str] = {},
+        rpc_timeout: Optional[timedelta] = None,
+    ) -> AsyncIterator[WorkflowHistory]:
+        """Create an async iterator consuming all workflows and calling
+        :py:meth:`WorkflowHandle.fetch_history` on each one.
+
+        This is just a shortcut for ``fetch_history``, see that method for
+        parameter details.
+        """
+        # TODO(cretz): Document that this consumes the iterator
```

Comment:
May as well just add that to the docstring?

Reply:
Just out of curiosity what happens if they iterate over the original iterator and this at the same time? Will it just dole out items to either the mapped history iterator or the original one in a racy way?

Reply:
Since it's an async iterator, it's whoever wins on each call. Basically the iterator is just calling `await my_iterator.__anext__()` on this class over and over until `StopAsyncIteration` is raised. The iterator doesn't care who's calling or, or if two things are calling it.

Reply:
Removed comment